esmvalcore/_recipe/from_datasets.py

```diff
@@ -0,0 +1,327 @@
+    _move_one_level_up(recipe, 'diagnostics', 'datasets')
+
+    return recipe
+
+
+def _move_one_level_up(base: dict, level: str, target: str):
+    """Move datasets one level up in the recipe."""
+    groups = base[level]
+
+    # Create a mapping from objects that can be hashed to the dicts
+    # describing the datasets.
+    dataset_mapping = {}
+    for name, group in groups.items():
+        dataset_mapping[name] = {
+            tuple(
+                sorted((k, tuple(v) if isinstance(v, list) else v)
+                       for k, v in ds.items())): ds
+            for ds in group['additional_datasets']
+        }
+
+    # Set datasets that are common to all groups
+    first_datasets = next(iter(dataset_mapping.values()))
+    common_datasets = set(first_datasets)
+    for datasets in dataset_mapping.values():
+        common_datasets &= set(datasets)
+    base[target] = [
+        v for k, v in first_datasets.items() if k in common_datasets
+    ]
+
+    # Remove common datasets from groups
+    for name, datasets in dataset_mapping.items():
+        group = groups[name]
+        var_datasets = set(datasets) - common_datasets
+        if var_datasets:
+            group['additional_datasets'] = [
+                v for k, v in datasets.items() if k in var_datasets
+            ]
+        else:
+            group.pop('additional_datasets')
+
+
+def _group_identical_facets(variable: Mapping[str, Any]) -> Recipe:
+    """Move identical facets from datasets to variable."""
+    result = dict(variable)
+    dataset_facets = result.pop('additional_datasets')
+    variable_keys = [
+        k for k, v in dataset_facets[0].items()
+        if k != 'dataset'  # keep at least one key in every dataset
+        and all((k, v) in d.items() for d in dataset_facets[1:])
+    ]
+    result.update(
+        (k, v) for k, v in dataset_facets[0].items() if k in variable_keys)
+    result['additional_datasets'] = [{
+        k: v
+        for k, v in d.items() if k not in variable_keys
+    } for d in dataset_facets]
+    return result
+
+
+def _group_ensemble_members(dataset_facets: Iterable[Facets]) -> list[Facets]:
+    """Group ensemble members.
+
+    This is the inverse operation of `Dataset.from_ranges` for
+    ensembles.
+    """
+
+    def grouper(facets):
+        return tuple((k, facets[k]) for k in sorted(facets) if k != 'ensemble')
+
+    result = []
+    for group_facets, group in itertools.groupby(dataset_facets, key=grouper):
+        ensembles = [f['ensemble'] for f in group if 'ensemble' in f]
+        if not ensembles:
+            result.append(dict(group_facets))
+        else:
+            for ensemble in _group_ensemble_names(ensembles):
+                facets = dict(group_facets)
+                facets['ensemble'] = ensemble
+                result.append(facets)
+    return result
+
+
+def _group_ensemble_names(ensemble_names: Iterable[str]) -> list[str]:
+    """Group ensemble names.
```

Comment:
It could be useful standalone, but you can already use it as part of the `datasets_to_recipe` function so maybe that is enough. I am not sure where it would fit in the API at the moment. Once we have a public `esmvalcore.recipe` module (possibly starting out from `esmvalcore.expirimental.recipe` it might fit in there..

Reply:
OK fine by me :+1: